compiler/util-klib-abi/src/org/jetbrains/kotlin/library/abi/parser/Cursor.kt

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright 2024 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.jetbrains.kotlin.library.abi.parser
+
+internal class Cursor
+private constructor(private val lines: List<String>, rowIndex: Int = 0, columnIndex: Int = 0) {
+    constructor(text: String) : this(text.split("\n"))
+
+    var rowIndex: Int = rowIndex
+        private set
+
+    var columnIndex: Int = columnIndex
+        private set
+
+    val currentLine: String
+        get() = lines[rowIndex].slice(columnIndex until lines[rowIndex].length)
+
+    val offset = lines.subList(0, rowIndex).sumOf { it.length } + columnIndex
+
+    /** Check if we have passed the last line in [lines] and there is nothing left to parse */
+    fun isFinished() = rowIndex >= lines.size
+
+    fun nextLine() {
+        rowIndex++
+        columnIndex = 0
+        if (!isFinished()) {
+            skipInlineWhitespace()
+        }
+    }
+
+    fun parseSymbol(
+        pattern: Regex,
+        peek: Boolean = false,
+        skipInlineWhitespace: Boolean = true,
+    ): String? {
+        val match = pattern.find(currentLine)
```

Comment:
`currentLine` creates a substring on every *call* and this function uses it twice. It makes sense to cache its value locally.